### DISCUSSION TITLE
Adding vike support for Purescript Deku

### DISCUSSION THREAD

**Discussion by mikesol** [vikejs/vike#1652] [Discussion]
Hi! I'm the lead developer of [PureScript Deku](https://purescript-deku.surge.sh/). I'm reading https://vike.dev/ui-framework and it recommends opening up an issue here to kick off the process of a vike integration. Thanks in advance for any resources or pointers!

**Comment by brillout** [comment]
Hi Mike, sounds interesting! What I recommend is to have a look at the minimal integrations `/examples/react-minimal` and `/examples/vue-minimal` (see root of this repo) and see if you can get a rough Vike + Deku example going. That alone would be lovely and we could create a `vike.dev/deku` page and link to that example.

Then, we can flesh out the example to something more full-fledged, see for example `boilerplates/boilerplate-*` (or `$ pnpm create vike`).

Once we have a satisfactory example going, we can create a Vike extension `vike-deku` or a framework that completely wraps Vike.

I guess the biggest question here is how to interop Vike with PureScript. Not sure if it's easily feasible with a friendly DX?

Feel free to let me know if you run into blockers.

**Reply by mikesol** [reply]
Sounds good, I'll put something together & ping you once it's done!

**Reply by brillout** [reply]
Neat, looking forward to it.

**Reply by mikesol** [reply]
Hello! Thanks for your patience. Life happened & I got sidetracked.

Here it is! https://github.com/mikesol/vite-deku-minimal/

Let me know if the installation and/or running instructions need polish. Hopefully the example works ok on your end. Once you take it for a spin, I'd love to hear your feedback.

To get it up and running, I needed to define a small `Util.purs` file with two functions, `toJSON` and `fromJSON`. All of the other functions imported into `+onRenderHtml.js` and `+onRenderClient.js` are from vanilla `deku`.

The DX for sure needs some love, as there are some warts, for example importing deku's page into its corresponding `+Page.js` document. I'd love to fix this up. Thanks in advance to the community for your help as I iterate this 🙏

**Reply by mikesol** [reply]
@brillout sorry I forgot to ping you in the last message (not sure if you're subscribed to this thread).

I added a small plugin that takes care of the codegen of `+Page.js`. So it's possible to just hack on the PureScript and new pages will appear automatically in the `pages` directory. There's lots of room for added options, for example a bit more type safety and configurable naming conventions, but this gets the ball over the net 🎾 

Vike currently skips `.gitignore` files, so the codegened stuff can't be ignored. But maybe that's fine?

Thanks in advance for your feedback!

**Reply by mikesol** [reply]
Added a client routing example as well: <https://github.com/mikesol/vike-deku-client-routing>.

**Reply by brillout** [reply]
Nice! Let me finish a couple of priorities and I'll have a look at it then.

**Reply by mikesol** [reply]
I migrated the deku docs to use vike + vercel & it worked like a charm!

- https://deku-documentation.vercel.app
- https://github.com/mikesol/deku-documentation

**Reply by mikesol** [reply]
Hi! I just wanted to touch base about this to see if it'd make sense to move forward with the integration. Thanks!

**Reply by brillout** [reply]
I just had a look at  https://github.com/mikesol/vike-deku-client-routing and it seems like you don't want to use Filesystem Routing? Because it isn't functional enough for your taste? I'm seeing `src/Vike/ClientRouting/Pages/Route.purs` and I wonder what concrete disadvantage you see in Filesystem Routing.

**Reply by mikesol** [reply]
Hi!
I'm definitely open to using filesystem routing.
The issue I ran into with it is that it created boilerplate, as the JS files only wound up importing Purescript logic. They could be auto-generated from the Purescript naming conventions at build time, but that seemed like a heavier lift than `Route.purs`.

**Reply by brillout** [reply]
:+1: Feel free to let me know if you need something on Vike's side, although note that we have quite a cramped roadmap already.

**Reply by mikesol** [reply]
All's good! If you think the repo is up to standards, it'd be nice to make an integration. I can of course use file system routing if you prefer 👍

**Reply by brillout** [reply]
I guess the best would be to create `vike-deku` under your GitHub name so that you're then free to go in whichever direction you feel like :) I'll happily list it in Vike's docs.
